### PR TITLE
refactor: simplify exponential backoff and refactor env

### PR DESCRIPTION
**Pull request by withinboredom**
This steals the movement of the `env` functions from #1137 and refactors the exponential backoff logic, while greatly simplifying it.

The new exponential backoff api is quite simple:

- `.wait()`: pauses execution if there has been a crash
- `.recordSuccess()`: resets the backoff state
- `.recordFailure()`: increments the pause and returns `true` if we've crashed too many times

### CODE REVIEW

File: backoff.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+type exponentialBackoff struct {
+	backoff                time.Duration
+	failureCount           int
+	mu                     sync.RWMutex
+	maxBackoff             time.Duration
+	minBackoff             time.Duration
+	maxConsecutiveFailures int
+}
+
+func newExponentialBackoff(minBackoff time.Duration, maxBackoff time.Duration, maxConsecutiveFailures int) *exponentialBackoff {
```
Comment (dunglas) [inline backoff.go:17]: nit: maybe can we remove this factory function, initializing the struct directly is clear and idiomatic.

File: backoff.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		maxConsecutiveFailures: maxConsecutiveFailures,
+	}
+}
+
+// recordSuccess resets the backoff and failureCount
+func (e *exponentialBackoff) recordSuccess() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.failureCount = 0
+	e.backoff = e.minBackoff
```
Comment (dunglas) [inline backoff.go:31]:
```suggestion
	e.failureCount = 0
	e.backoff = e.minBackoff
	e.mu.Unlock()
```

`defer` may cost a bit (I didn't check in this specific case): https://victoriametrics.com/blog/defer-in-go/

Here the code is so simple that it's maybe not necessary to use defer?

File: backoff.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+// recordFailure increments the failure count and increases the backoff, it returns true if maxConsecutiveFailures has been reached
+func (e *exponentialBackoff) recordFailure() bool {
+	e.mu.Lock()
+	defer e.mu.Unlock()
```
Comment (dunglas) [inline backoff.go:37]: same here maybe? especially after having applied the linter suggestion

File: backoff.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+	return false
+}
+
+// wait sleeps for the backoff duration if failureCount is non-zero.
+// NOTE: this is not tested and should be kept 'obviously correct' (i.e., simple)
+func (e *exponentialBackoff) wait() {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	if e.failureCount == 0 {
```
Comment (dunglas) [inline backoff.go:53]: IIUC, you should unlock after the read but before wating, right?

File: backoff_test.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	e.recordSuccess()
+
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	if e.failureCount != 0 {
+		t.Errorf("expected failureCount to be reset to 0, got %d", e.failureCount)
+	}
+	if e.backoff != e.minBackoff {
+		t.Errorf("expected backoff to be reset to minBackoff, got %v", e.backoff)
+	}
```
Comment (dunglas) [inline backoff_test.go:23]:
```suggestion
	assert.Equal(t, 0, e.failureCount, "expected failureCount to be reset to 0")
	assert.Equal(t, e.minBackoff, e.backoff, "expected backoff to be reset to minBackoff")
```

File: backoff_test.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	assert.True(t, e.recordFailure())
+
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	if e.failureCount != e.maxConsecutiveFailures {
+		t.Errorf("expected failureCount to be %d, got %d", e.maxConsecutiveFailures, e.failureCount)
+	}
+	if e.backoff != e.maxBackoff {
+		t.Errorf("expected backoff to be maxBackoff, got %v", e.backoff)
+	}
```
Comment (dunglas) [inline backoff_test.go:40]: same here, you can use `assert.Equal()`

File: env.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	"unsafe"
+)
+
+//export go_putenv
+func go_putenv(str *C.char, length C.int) C.bool {
+	// Create a byte slice from C string with a specified length
+	s := C.GoBytes(unsafe.Pointer(str), length)
+
+	// Convert byte slice to string
+	envString := string(s)
```
Comment (dunglas) [inline env.go:17]:
```suggestion
	envString := C.GoStringN(str, length)
```

File: env.go [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+			return false // Failure
+		}
+	} else {
+		// No '=', unset the environment variable
+		if os.Unsetenv(envString) != nil {
+			return false // Failure
+		}
+	}
+
+	return true // Success
```
Comment (dunglas) [inline env.go:31]:
```suggestion
	// Check if '=' is present in the string
	if key, val, found := strings.Cut(envString, "="); found {
		return os.Setenv(key, val) == nil
	}

	// No '=', unset the environment variable
	return os.Unsetenv(envString) == nil
```

**Review by dunglas — COMMENTED** [review]
Good refactoring!

File: backoff.go (hunk 1 above)
Comment (withinboredom) [reply backoff.go]: Nice one.

**Comment by AlliBalliBaba** [comment]
As a sidenote, the segfault in the 8.3 tests also randomly happens on main, see #1141

**Review by dunglas — APPROVED** [review]

**Pull request merged**